plugin/cache/cache_test.go

```diff
@@ -275,74 +43,303 @@ func newTestCache(ttl time.Duration) (*Cache, *ResponseWriter) {
+				Qname: "neg-disabled.example.org.", Qtype: dns.TypeA,
+				Answer: []dns.RR{
+					test.A("neg-disabled.example.org. 3600 IN	A	127.0.0.1"),
+				},
+			},
+			shouldCache: true,
+		},
+	}
 	now, _ := time.Parse(time.UnixDate, "Fri Apr 21 10:51:21 BST 2017")
 	utc := now.UTC()
 
-	for n, tc := range cacheTestCases {
-		// Create a new cache every time to prevent accidental comparison with a previous item.
-		c, crr := newTestCache(maxTTL)
-
-		m := tc.in.Msg()
-		m = cacheMsg(m, tc.in)
-
-		state := request.Request{W: &test.ResponseWriter{}, Req: m}
+	for _, tc := range cacheTestCases {
+		t.Run(tc.name, func(t *testing.T) {
+			// Create a new cache every time to prevent accidental comparison with a previous item.
+			c, crr := newTestCache(maxTTL)
 
-		mt, _ := response.Typify(m, utc)
-		valid, k := key(state.Name(), m, mt, state.Do(), state.Req.CheckingDisabled)
+			m := tc.in.Msg()
+			m = cacheMsg(m, tc.in)
 
-		if valid {
-			// Insert cache entry
-			crr.set(m, k, mt, c.pttl)
-		}
+			state := request.Request{W: &test.ResponseWriter{}, Req: m}
 
-		// Attempt to retrieve cache entry
-		i := c.getIgnoreTTL(time.Now().UTC(), state, "dns://:53")
-		found := i != nil
+			mt, _ := response.Typify(m, utc)
+			valid, k := key(state.Name(), m, mt, state.Do(), state.Req.CheckingDisabled)
 
-		if !tc.shouldCache && found {
-			t.Errorf("Test %d: Cached message that should not have been cached: %s", n, state.Name())
-			continue
-		}
-		if tc.shouldCache && !found {
-			t.Errorf("Test %d: Did not cache message that should have been cached: %s", n, state.Name())
-			continue
-		}
+			if valid {
+				// Insert cache entry
+				crr.set(m, k, mt, c.pttl)
+			}
 
-		if found {
-			resp := i.toMsg(m, time.Now().UTC(), state.Do(), m.AuthenticatedData)
+			// Attempt to retrieve cache entry
+			i := c.getIgnoreTTL(time.Now().UTC(), state, "dns://:53")
+			found := i != nil
 
-			// TODO: If we incorporate these individual checks into the
-			//       test.Header function, we can eliminate them from here.
-			// Cache entries are always Authoritative.
-			if resp.Authoritative != true {
-				t.Error("Expected Authoritative Answer bit to be true, but was false")
-			}
-			if resp.AuthenticatedData != tc.out.AuthenticatedData {
-				t.Errorf("Expected Authenticated Data bit to be %t, but got %t", tc.out.AuthenticatedData, resp.AuthenticatedData)
+			if !tc.shouldCache && found {
+				t.Errorf("Cached message that should not have been cached: %s", state.Name())
 			}
-			if resp.RecursionAvailable != tc.out.RecursionAvailable {
-				t.Errorf("Expected Recursion Available bit to be %t, but got %t", tc.out.RecursionAvailable, resp.RecursionAvailable)
-			}
-			if resp.CheckingDisabled != tc.out.CheckingDisabled {
-				t.Errorf("Expected Checking Disabled bit to be %t, but got %t", tc.out.CheckingDisabled, resp.CheckingDisabled)
+			if tc.shouldCache && !found {
+				t.Errorf("Did not cache message that should have been cached: %s", state.Name())
```

Comment:
These should be `t.Fatalf` now.  You still want to go to the next test case (which pre-`t.Run`-refactoring was done by `continue`) rather than do all the `if resp.foo != tc.out.foo { t.Errorf(...) }` checks below if you didn't find a cache entry or got an unexpected cache entry.  (Alternatively, you could make the `if found` into an `else if found`.)

Reply:
Done - I debated internally about this and though it wouldn't hurt to continue checking the improperly found cache item, but now that seems pointless since the test is already invalid.